### ISSUE TITLE
Links to media is broken when accessing Tor via hidden service since v4

### ISSUE THREAD

**Issue by aphick** [mastodon/mastodon#22085]
### Steps to reproduce the problem

Assets such as headers, avatars, and uploads are served through the https:// + LOCAL_DOMAIN address rather than over the Tor hidden service and http. This issue is not present in v3.5.5

### Expected behaviour

Serve assets through relative links

### Actual behaviour

Statically linked assets are served

### Specifications

Mastodon | 4.0.2
Ruby | 3.0.4p208
PostgreSQL | 15.1
Redis | 6.0.16

**Comment by rbairwell** [comment]
Is the configuration variable `[CDN_HOST](https://docs.joinmastodon.org/admin/config/#cdn_host)` and [S3_ALIAS_HOST ](https://docs.joinmastodon.org/admin/config/#s3_alias_host) configuration settings set to point to the correct TOR endpoints?

**Comment by aphick** [comment]
@rbairwell No it was not. That did the trick Tor only access to the instance, but would not work outside Tor correct?

**Comment by aphick** [comment]
Support for a Tor only instance would be great, however, I've found another issue while testing federation from an instance configured to use privoxy and with `http_proxy` set connecting to an instance using a .onion HS as its LOCAL_DOMAIN. The Tor HS domain links to `https://tordomain.onion` in /.well-known/host-meta, so federation would not work in a purely Tor use-case.

**Comment by rbairwell** [comment]
Okay, a little bit of digging shows nginx has the ['ngx_http_sub_module'](https://nginx.org/en/docs/http/ngx_http_sub_module.html) which would allow instructions such as (untested):
```
server {
  listen 80;
  server_name instance.onion
  location / {
      sub_filter 'https://image-server.example.com/' 'http://image-server.onion/';
      sub_filter 'https://instance.example.com/' 'http://instance.onion/';
      sub_filter_types text/html text/plain text/css application/json application/javascript text/xml application/xml application/xml+rss text/javascript application/activity+json
  }
}

```
change any non-onion links to your servers in html, css, json, javascript and xml files to the onion one (but only on the `instance.onion` server as [defined in the docs](https://docs.joinmastodon.org/admin/optional/tor/#http).

(There is also nginx's [subsititutions_filter ](https://www.nginx.com/resources/wiki/modules/substitutions/) which does a similar job).

Be warned, since nginx will have to parse/scan all of those files, there will be an overhead to it. It also might not work for "streamed" content (such as updates) as that is served through a separate server ( [streaming/node.js](https://github.com/mastodon/mastodon/blob/main/streaming/index.js) )

Can you try that change and see if it helps/works at all? I haven't got (at the moment) full access to a Mastodon server I can test it on (and, tbh, neither an nginx server at the moment!) and I don't have TOR setup on any servers. If it does, I'll make a PR for the documentation for other users and then have a look to see if I can do anything with the streaming server.

**Comment by aphick** [comment]
Yes @rbairwell I was trying out sub_filter a little yesterday, and I was able to get the Tor instance admin account to show up in search and to send off a follow request, however federation was never fully completed and I see that the clearnet instance is still trying to access over https://

```
